magenta-lib/src/main/scala/magenta/deployment_type/Param.scala

```diff
@@ -38,21 +40,24 @@ case class Param[T](
       (DeploymentPackage, DeployTarget) => Either[String, T]
     ] = None,
     deprecatedDefault: Boolean = false
-)(implicit register: ParamRegister) {
+)(implicit register: ParamRegister, reads: Reads[T], manifest: Manifest[T]) {
   register.add(this)
 
   val required =
     !optional && defaultValue.isEmpty && defaultValueFromContext.isEmpty
 
-  def get(pkg: DeploymentPackage)(implicit reads: Reads[T]): Option[T] =
+  def get(pkg: DeploymentPackage): Option[T] =
     pkg.pkgSpecificData
       .get(name)
-      .flatMap(jsValue => Json.fromJson[T](jsValue).asOpt)
+      .flatMap(jsValue => parse(jsValue))
+
+  def parse(jsValue: JsValue): Option[T] = Json.fromJson[T](jsValue).asOpt
+
   def apply(
       pkg: DeploymentPackage,
       target: DeployTarget,
       reporter: DeployReporter
-  )(implicit reads: Reads[T], manifest: Manifest[T]): T = {
+  ): T = {
```

Comment:
This method does need a `Reads[T]` instance, to know how to parse a JSON value into an instance of a `T`, but asking for it to be supplied here, where the method is called, means that every single caller of this code must have the _right_ `Reads[T]` instance available, and they must all agree, otherwise a parameter may get parsed differently depending on which code calls it! This makes testing that a parameter is being correctly parsed harder than it needs to be.

Better to add the requirement for the `Reads[T]` to the _constructor_ of the `Param[T]` class, so that once the `Param[T]` is constructed, it always behave in the same way. Consequently, we can write a test using a param - like the `healthcheckGrace` param - and be sure that it will behave the same way when the implementation code calls it.